File: apps/backend/src/routes/enhancedMemoryRoutes.js
/**
 * Enhanced Memory Routes - Express.js REST API routes for Graphiti Knowledge Graph Integration
 * 
 * Note: Converting from FastAPI/Python format to Express.js/Node.js format
 * for compatibility with the existing SUPERmcp architecture
 */

const express = require('express');
const router = express.Router();
const { body, query, validationResult } = require('express-validator');
const logger = require('../utils/logger');

// Mock controller for now - will be implemented in Python service
class EnhancedMemoryController {
    async storeMemory(req, res, next) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    error: 'Validation failed',
                    details: errors.array()
                });
            }

            // This will call the Python service via HTTP or direct Python integration
            // For now, return a mock response indicating the service is being implemented
            const response = {
                success: true,
                message: "Enhanced Memory System with Graphiti integration is being implemented",
                memory_id: `mem_${Date.now()}`,
                graph_episode_id: `graph_${Date.now()}`,
                relationships_found: 0,
                entities_found: 0,
                note: "This endpoint will integrate with the Python-based Graphiti service"
            };

            logger.info('📝 Enhanced memory store request received', {
                content_length: req.body.content?.length || 0,
                has_context: !!req.body.context,
                agent_name: req.body.agent_name
            });

            res.json(response);

        } catch (error) {
            logger.error('Enhanced memory store failed', { error: error.message });
            next(error);
        }
    }

    async searchMemory(req, res, next) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    error: 'Validation failed',
                    details: errors.array()
                });
            }

            // Mock response for enhanced search
            const response = {
                success: true,
                results: [
                    {
                        content: "Sample enhanced memory result with graph relationships",
                        relevance_score: 0.85,
                        source: "hybrid",
                        relationships: ["entity1 -> relates_to -> entity2"],
                        metadata: { type: "sample" }
                    }
                ],
                total_results: 1,
                search_type: req.body.search_type || "hybrid",
                sources_used: ["vector", "graph"],
                query_time_ms: 150,
                note: "Enhanced search combining vector similarity and graph traversal"
            };

            logger.info('🔍 Enhanced memory search request received', {
                query: req.body.query?.substring(0, 50) + '...',
                search_type: req.body.search_type,
                limit: req.body.limit
            });

            res.json(response);

        } catch (error) {
            logger.error('Enhanced memory search failed', { error: error.message });
            next(error);
        }
    }

    async temporalSearch(req, res, next) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    error: 'Validation failed',
                    details: errors.array()
                });
            }

            const response = {
                success: true,
                results: [
                    {
                        content: "Temporal search result showing relationship evolution",
                        relevance_score: 0.78,
                        temporal_context: {
                            evolution: "Relationship strengthened over time",
                            time_period: "last_30_days"
                        }
                    }
                ],
                temporal_insights: {
                    patterns_detected: ["strengthening_relationships", "emerging_clusters"],
                    evolution_summary: "Knowledge graph has grown 25% in complexity"
                },
                relationship_evolution: [],
                patterns_detected: ["temporal_clustering", "relationship_strengthening"]
            };

            logger.info('⏰ Temporal search request received', {
                query: req.body.query?.substring(0, 50),
                time_range: req.body.start_time && req.body.end_time ? 'specified' : 'all_time'
            });

            res.json(response);

        } catch (error) {
            logger.error('Temporal search failed', { error: error.message });
            next(error);
        }
    }

    async getAgentCollaboration(req, res, next) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    error: 'Validation failed',
                    details: errors.array()
                });
            }

            const response = {
                success: true,
                agent_context: {
                    agent_name: req.body.agent_name,
                    capabilities: ["analysis", "reasoning", "memory_management"],
                    current_knowledge: "Domain expertise in the requested area"
                },
                collaboration_opportunities: [
                    {
                        agent: "manus",
                        opportunity: "Share task execution patterns",
                        confidence: 0.85
                    },
                    {
                        agent: "sam",
                        opportunity: "Exchange analytical insights",
                        confidence: 0.92
                    }
                ],
                shared_knowledge: {
                    common_patterns: ["task_completion", "user_preferences"],
                    collaboration_history: "15 successful interactions"
                }
            };

            logger.info('🤝 Agent collaboration request received', {
                agent_name: req.body.agent_name,
                task_id: req.body.task_id
            });

            res.json(response);

        } catch (error) {
            logger.error('Agent collaboration failed', { error: error.message });
            next(error);
        }
    }

    async analyzeKnowledgeGaps(req, res, next) {
        try {
            const errors = validationResult(req);
            if (!errors.isEmpty()) {
                return res.status(400).json({
                    success: false,
                    error: 'Validation failed',
                    details: errors.array()
                });
            }

            const response = {
                success: true,
                analysis: {
                    missing_relationships: [
                        {
                            entities: ["concept_a", "concept_b"],
                            potential_relationship: "depends_on",
                            confidence: 0.75
                        }
                    ],
                    weak_connections: [
                        {
                            domain: "task_management",
                            strength: 0.3,
                            improvement_potential: "high"
                        }
                    ],
                    enrichment_opportunities: [
                        "Add more contextual relationships",
                        "Strengthen temporal connections",
                        "Improve entity resolution"
                    ]
                },
                recommendations: [
                    {
                        type: "missing_relationship",
                        priority: "high",
                        action: "Establish connection between related concepts",
                        expected_impact: "Improved contextual understanding"
                    }
                ],
                priority_areas: ["relationship_discovery", "knowledge_enrichment"]
            };

            logger.info('🔍 Knowledge gap analysis request received', {
                query: req.body.query?.substring(0, 50)
            });

            res.json(response);

        } catch (error) {
            logger.error('Knowledge gap analysis failed', { error: error.message });
            next(error);
        }
    }

    async predictNextActions(req, res, next) {
        try {
            const response = {
                success: true,
                predictions: [
                    {
                        suggested_action: "Review related documents in the knowledge graph",
                        confidence: 0.82,
                        reasoning: "Based on similar interaction patterns",
                        prediction_type: "next_task",
                        based_on_patterns: ["document_review_sequence", "user_behavior"]
                    },
                    {
                        suggested_action: "Analyze temporal relationships for this topic",
                        confidence: 0.76,
                        reasoning: "Historical data suggests temporal analysis follows this query type",
                        prediction_type: "analytical_insight",
                        based_on_patterns: ["temporal_analysis_pattern"]
                    }
                ],
                total_predictions: 2,
                note: "Predictions based on knowledge graph patterns and user behavior analysis"
            };

            logger.info('🔮 Predictive analysis request received', {
                query: req.query.query?.substring(0, 50)
            });

            res.json(response);

        } catch (error) {
            logger.error('Predictive analysis failed', { error: error.message });
            next(error);
        }
    }

    async getSystemMetrics(req, res, next) {
        try {
            const response = {
                success: true,
                system_metrics: {
                    total_episodes: 0,
                    successful_stores: 0,
                    failed_stores: 0,
                    search_operations: 0,
                    graph_operations: 0,
                    relationships_created: 0,
                    predictions_made: 0,
                    system_status: {
                        dual_write_enabled: true,
                        neo4j_status: "not_connected",
                        graphiti_status: "initializing",
                        temporal_reasoning_enabled: true,
                        predictive_assistance_enabled: true,
                        fallback_to_vector: true
                    }
                },
                controller_metrics: {
                    controller_requests: 0,
                    controller_errors: 0,
                    error_rate: 0
                },
                timestamp: new Date().toISOString(),
                note: "Graphiti integration is being implemented - metrics will be available once Python service is connected"
            };

            res.json(response);

        } catch (error) {
            logger.error('System metrics failed', { error: error.message });
            next(error);
        }
    }

    async healthCheck(req, res, next) {
        try {
            const response = {
                overall_status: "initializing",
                components: {
                    vector_memory: "healthy",
                    neo4j: "not_configured",
                    graphiti: "initializing",
                    enhanced_memory_controller: "ready"
                },
                metrics: {
                    total_operations: 0,
                    success_rate: "100%",
                    avg_latency_ms: 0
                },
                timestamp: new Date().toISOString(),
                message: "Enhanced Memory System is being integrated with Graphiti knowledge graph"
            };

            res.json(response);

        } catch (error) {
            logger.error('Health check failed', { error: error.message });
            res.status(500).json({
                overall_status: "error",
                components: { health_check: `error: ${error.message}` },
                metrics: {},
                timestamp: new Date().toISOString()
            });
        }
    }

    async getSearchTypes(req, res, next) {
        try {
            const response = {
                search_types: {
                    VECTOR_ONLY: {
                        description: "Traditional vector similarity search using embeddings",
                        use_cases: ["Simple content retrieval", "Fast searches", "Fallback mode"],
                        performance: "Fast",
                        accuracy: "Good"
                    },
                    GRAPH_ONLY: {
                        description: "Knowledge graph traversal search using relationships",
                        use_cases: ["Relationship exploration", "Context discovery", "Entity connections"],
                        performance: "Medium",
                        accuracy: "High contextual"
                    },
                    HYBRID: {
                        description: "Combined vector + graph search with intelligent merging",
                        use_cases: ["Best overall results", "Complex queries", "Contextual search"],
                        performance: "Medium",
                        accuracy: "Highest"
                    },
                    TEMPORAL: {
                        description: "Time-aware search with relationship evolution",
                        use_cases: ["Historical analysis", "Change tracking", "Temporal patterns"],
                        performance: "Slower",
                        accuracy: "High temporal"
                    },
                    PREDICTIVE: {
                        description: "Pattern-based search with future predictions",
                        use_cases: ["Proactive assistance", "Pattern recognition", "Trend analysis"],
                        performance: "Slower",
                        accuracy: "Predictive"
                    }
                },
                recommendations: {
                    general_use: "HYBRID",
                    fast_retrieval: "VECTOR_ONLY",
                    relationship_exploration: "GRAPH_ONLY",
                    historical_analysis: "TEMPORAL",
                    proactive_assistance: "PREDICTIVE"
                }
            };

            res.json(response);

        } catch (error) {
            logger.error('Get search types failed', { error: error.message });
            next(error);
        }
    }

    async explainSearchResults(req, res, next) {
        try {
            const { query, search_type = 'HYBRID' } = req.body;

            const response = {
                query,
                search_type,
                strategy: {
                    description: `Used ${search_type} search strategy`,
                    components: [],
                    ranking_factors: []
                },
                results_analysis: {
                    total_results: 1,
                    sources_used: ["vector", "graph"],
                    avg_relevance: 0.85
                },
                performance: {
                    query_time_ms: 150,
                    efficiency_rating: "High"
                },
                note: "Search explanation will provide detailed insights once Graphiti integration is complete"
            };

            // Add strategy-specific details
            if (['VECTOR_ONLY', 'HYBRID'].includes(search_type)) {
                response.strategy.components.push("Vector similarity using embeddings");
                response.strategy.ranking_factors.push("Cosine similarity score");
            }

            if (['GRAPH_ONLY', 'HYBRID'].includes(search_type)) {
                response.strategy.components.push("Knowledge graph traversal");
                response.strategy.ranking_factors.push("Relationship strength");
            }

            if (search_type === 'TEMPORAL') {
                response.strategy.components.push("Temporal reasoning");
                response.strategy.ranking_factors.push("Temporal relevance");
            }

            if (search_type === 'PREDICTIVE') {
                response.strategy.components.push("Pattern analysis");
                response.strategy.ranking_factors.push("Prediction confidence");
            }

            logger.info('📊 Search explanation request received', {
                query: query?.substring(0, 50),
                search_type
            });

            res.json(response);

        } catch (error) {
            logger.error('Search explanation failed', { error: error.message });
            next(error);
        }
    }
}

// Create controller instance
const controller = new EnhancedMemoryController();

// Validation middleware
const validateStoreMemory = [
    body('content').notEmpty().withMessage('Content is required'),
    body('context').optional().isObject().withMessage('Context must be an object'),
    body('agent_name').optional().isString().withMessage('Agent name must be a string'),
    body('task_id').optional().isString().withMessage('Task ID must be a string')
];

const validateSearchMemory = [
    body('query').notEmpty().withMessage('Query is required'),
    body('search_type').optional().isIn(['VECTOR_ONLY', 'GRAPH_ONLY', 'HYBRID', 'TEMPORAL', 'PREDICTIVE']).withMessage('Invalid search type'),
    body('user_context').optional().isObject().withMessage('User context must be an object'),
    body('limit').optional().isInt({ min: 1, max: 100 }).withMessage('Limit must be between 1 and 100'),
    body('start_time').optional().isISO8601().withMessage('Start time must be a valid ISO date'),
    body('end_time').optional().isISO8601().withMessage('End time must be a valid ISO date')
];

const validateTemporalSearch = [
    body('query').notEmpty().withMessage('Query is required'),
    body('start_time').optional().isISO8601().withMessage('Start time must be a valid ISO date'),
    body('end_time').optional().isISO8601().withMessage('End time must be a valid ISO date'),
    body('include_evolution').optional().isBoolean().withMessage('Include evolution must be boolean'),
    body('limit').optional().isInt({ min: 1, max: 50 }).withMessage('Limit must be between 1 and 50')
];

const validateAgentCollaboration = [
    body('agent_name').notEmpty().withMessage('Agent name is required'),
    body('task_id').optional().isString().withMessage('Task ID must be a string'),
    body('include_shared_knowledge').optional().isBoolean().withMessage('Include shared knowledge must be boolean')
];

const validateKnowledgeGaps = [
    body('query').notEmpty().withMessage('Query is required'),
    body('context').optional().isObject().withMessage('Context must be an object')
];

const validateSearchExplanation = [
    body('query').notEmpty().withMessage('Query is required'),
    body('search_type').optional().isIn(['VECTOR_ONLY', 'GRAPH_ONLY', 'HYBRID', 'TEMPORAL', 'PREDICTIVE']).withMessage('Invalid search type')
];

// Routes

/**
 * POST /api/memory/enhanced/store
 * Store content in enhanced memory system
 */
router.post('/store', validateStoreMemory, controller.storeMemory);

/**
 * POST /api/memory/enhanced/search
 * Search memory with hybrid approach
 */
router.post('/search', validateSearchMemory, controller.searchMemory);

/**
 * POST /api/memory/enhanced/temporal
 * Temporal reasoning queries
 */
router.post('/temporal', validateTemporalSearch, controller.temporalSearch);

/**
 * POST /api/memory/enhanced/collaboration
 * Agent collaboration context
 */
router.post('/collaboration', validateAgentCollaboration, controller.getAgentCollaboration);

/**
 * POST /api/memory/enhanced/gaps
 * Knowledge gap analysis
 */
router.post('/gaps', validateKnowledgeGaps, controller.analyzeKnowledgeGaps);

/**
 * POST /api/memory/enhanced/predict
 * Predictive assistance
 */
router.post('/predict', 
    query('query').notEmpty().withMessage('Query parameter is required'),
    controller.predictNextActions
);

/**
 * GET /api/memory/enhanced/metrics
 * System metrics
 */
router.get('/metrics', controller.getSystemMetrics);

/**
 * GET /api/memory/enhanced/health
 * Health check
 */
router.get('/health', controller.healthCheck);

/**
 * GET /api/memory/enhanced/search/types
 * Available search types
 */
router.get('/search/types', controller.getSearchTypes);

/**
 * POST /api/memory/enhanced/search/explain
 * Explain search results
 */
router.post('/search/explain', validateSearchExplanation, controller.explainSearchResults);

module.exports = router;

File: apps/backend/src/routes/enhancedMemoryRoutes.py
"""
Enhanced Memory Routes - REST API routes for Graphiti Knowledge Graph Integration

Provides comprehensive routing for:
- Hybrid memory storage and retrieval
- Temporal reasoning queries
- Agent collaboration features
- Predictive assistance
- Knowledge gap analysis
- System monitoring
"""

from fastapi import APIRouter, HTTPException, Depends
from typing import Dict, List, Optional, Any
import os
import json

from ..controllers.enhancedMemoryController import (
    EnhancedMemoryController,
    StoreMemoryRequest,
    SearchMemoryRequest,
    TemporalSearchRequest,
    AgentCollaborationRequest,
    KnowledgeGapRequest,
    StoreMemoryResponse,
    SearchMemoryResponse,
    TemporalSearchResponse,
    AgentCollaborationResponse,
    HealthCheckResponse,
    get_enhanced_memory_controller
)
from ..services.enhancedMemoryService import SearchType
from ..middleware.authMiddleware import AuthMiddleware
from ..utils.logger import logger

# Create router
router = APIRouter(prefix="/api/memory/enhanced", tags=["Enhanced Memory"])

# Auth middleware
auth_middleware = AuthMiddleware()

def get_controller() -> EnhancedMemoryController:
    """Dependency to get enhanced memory controller"""
    config = {
        'NEO4J_URI': os.getenv('NEO4J_URI', 'bolt://localhost:7687'),
        'NEO4J_USERNAME': os.getenv('NEO4J_USERNAME', 'neo4j'),
        'NEO4J_PASSWORD': os.getenv('NEO4J_PASSWORD', 'neo4j_password'),
        'OPENAI_API_KEY': os.getenv('OPENAI_API_KEY'),
        'GRAPHITI_EMBEDDING_MODEL': os.getenv('GRAPHITI_EMBEDDING_MODEL', 'text-embedding-3-small'),
        'GRAPHITI_LLM_MODEL': os.getenv('GRAPHITI_LLM_MODEL', 'gpt-4-turbo-preview')
    }
    return get_enhanced_memory_controller(config)

@router.post(
    "/store",
    response_model=StoreMemoryResponse,
    summary="Store content in enhanced memory system",
    description="Store content in both vector database and knowledge graph with relationship extraction"
)
async def store_memory(
    request: StoreMemoryRequest,
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)  # Uncomment when auth is ready
) -> StoreMemoryResponse:
    """
    Store content in the enhanced memory system with dual-write capability.
    
    Features:
    - Stores in both vector database and knowledge graph
    - Extracts entities and relationships automatically
    - Supports contextual metadata
    - Zero-downtime migration strategy
    """
    try:
        return await controller.store_memory(request)
    except Exception as error:
        logger.error(f"Store memory endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

@router.post(
    "/search",
    response_model=SearchMemoryResponse,
    summary="Search memory with hybrid approach",
    description="Search using combination of vector similarity and graph traversal"
)
async def search_memory(
    request: SearchMemoryRequest,
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)
) -> SearchMemoryResponse:
    """
    Search memory using advanced hybrid approach.
    
    Search Types:
    - VECTOR_ONLY: Traditional vector similarity search
    - GRAPH_ONLY: Knowledge graph traversal search
    - HYBRID: Combined vector + graph search (recommended)
    - TEMPORAL: Time-aware search with relationship evolution
    - PREDICTIVE: Pattern-based predictive search
    """
    try:
        return await controller.search_memory(request)
    except Exception as error:
        logger.error(f"Search memory endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

@router.post(
    "/temporal",
    response_model=TemporalSearchResponse,
    summary="Temporal reasoning queries",
    description="Perform time-aware searches with relationship evolution analysis"
)
async def temporal_search(
    request: TemporalSearchRequest,
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)
) -> TemporalSearchResponse:
    """
    Perform temporal reasoning queries to understand how knowledge evolved over time.
    
    Features:
    - Point-in-time queries
    - Relationship evolution tracking
    - Temporal pattern detection
    - Historical context analysis
    """
    try:
        return await controller.temporal_search(request)
    except Exception as error:
        logger.error(f"Temporal search endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

@router.post(
    "/collaboration",
    response_model=AgentCollaborationResponse,
    summary="Agent collaboration context",
    description="Get contextual knowledge for agent collaboration and shared intelligence"
)
async def get_agent_collaboration(
    request: AgentCollaborationRequest,
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)
) -> AgentCollaborationResponse:
    """
    Get agent collaboration context and opportunities.
    
    Features:
    - Agent-specific knowledge context
    - Collaboration opportunity detection
    - Shared knowledge graph analysis
    - Cross-agent learning insights
    """
    try:
        return await controller.get_agent_collaboration(request)
    except Exception as error:
        logger.error(f"Agent collaboration endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

@router.post(
    "/gaps",
    summary="Knowledge gap analysis",
    description="Analyze knowledge gaps and suggest improvements"
)
async def analyze_knowledge_gaps(
    request: KnowledgeGapRequest,
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)
) -> Dict[str, Any]:
    """
    Analyze knowledge gaps in the system and provide recommendations.
    
    Features:
    - Missing relationship detection
    - Weak connection identification
    - Knowledge enrichment suggestions
    - Priority area identification
    """
    try:
        return await controller.analyze_knowledge_gaps(request)
    except Exception as error:
        logger.error(f"Knowledge gap analysis endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

@router.post(
    "/predict",
    summary="Predictive assistance",
    description="Predict next actions based on knowledge graph patterns"
)
async def predict_next_actions(
    query: str,
    user_context: Optional[Dict[str, Any]] = None,
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)
) -> Dict[str, Any]:
    """
    Predict what user might need next based on graph patterns and historical behavior.
    
    Features:
    - Pattern-based predictions
    - Historical behavior analysis
    - Confidence scoring
    - Actionable recommendations
    """
    try:
        return await controller.predict_next_actions(query, user_context)
    except Exception as error:
        logger.error(f"Predictive assistance endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

@router.get(
    "/metrics",
    summary="System metrics",
    description="Get comprehensive system metrics and performance data"
)
async def get_system_metrics(
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)
) -> Dict[str, Any]:
    """
    Get comprehensive system metrics and performance data.
    
    Includes:
    - Storage statistics
    - Search performance
    - Graph operations
    - Relationship metrics
    - Error rates
    """
    try:
        return await controller.get_system_metrics()
    except Exception as error:
        logger.error(f"System metrics endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

@router.get(
    "/health",
    response_model=HealthCheckResponse,
    summary="Health check",
    description="Comprehensive health check for all system components"
)
async def health_check(
    controller: EnhancedMemoryController = Depends(get_controller)
) -> HealthCheckResponse:
    """
    Perform comprehensive health check of all system components.
    
    Checks:
    - Vector memory system
    - Neo4j connection
    - Graphiti functionality
    - System metrics
    """
    try:
        return await controller.health_check()
    except Exception as error:
        logger.error(f"Health check endpoint failed: {error}")
        return HealthCheckResponse(
            overall_status="error",
            components={"health_check": f"error: {error}"},
            metrics={},
            timestamp="error"
        )

# Advanced search endpoints

@router.get(
    "/search/types",
    summary="Available search types",
    description="Get available search types and their descriptions"
)
async def get_search_types() -> Dict[str, Any]:
    """Get available search types and their descriptions"""
    return {
        "search_types": {
            "VECTOR_ONLY": {
                "description": "Traditional vector similarity search using embeddings",
                "use_cases": ["Simple content retrieval", "Fast searches", "Fallback mode"],
                "performance": "Fast",
                "accuracy": "Good"
            },
            "GRAPH_ONLY": {
                "description": "Knowledge graph traversal search using relationships",
                "use_cases": ["Relationship exploration", "Context discovery", "Entity connections"],
                "performance": "Medium",
                "accuracy": "High contextual"
            },
            "HYBRID": {
                "description": "Combined vector + graph search with intelligent merging",
                "use_cases": ["Best overall results", "Complex queries", "Contextual search"],
                "performance": "Medium",
                "accuracy": "Highest"
            },
            "TEMPORAL": {
                "description": "Time-aware search with relationship evolution",
                "use_cases": ["Historical analysis", "Change tracking", "Temporal patterns"],
                "performance": "Slower",
                "accuracy": "High temporal"
            },
            "PREDICTIVE": {
                "description": "Pattern-based search with future predictions",
                "use_cases": ["Proactive assistance", "Pattern recognition", "Trend analysis"],
                "performance": "Slower",
                "accuracy": "Predictive"
            }
        },
        "recommendations": {
            "general_use": "HYBRID",
            "fast_retrieval": "VECTOR_ONLY",
            "relationship_exploration": "GRAPH_ONLY",
            "historical_analysis": "TEMPORAL",
            "proactive_assistance": "PREDICTIVE"
        }
    }

@router.post(
    "/search/explain",
    summary="Explain search results",
    description="Get detailed explanation of how search results were derived"
)
async def explain_search_results(
    query: str,
    search_type: SearchType = SearchType.HYBRID,
    controller: EnhancedMemoryController = Depends(get_controller),
    # current_user = Depends(auth_middleware.get_current_user)
) -> Dict[str, Any]:
    """
    Get detailed explanation of how search results were derived.
    
    Provides insight into:
    - Search strategy used
    - Vector similarity scores
    - Graph traversal paths
    - Ranking algorithm details
    - Performance metrics
    """
    try:
        # Perform search with detailed logging
        search_request = SearchMemoryRequest(
            query=query,
            search_type=search_type,
            limit=5
        )
        
        results = await controller.search_memory(search_request)
        
        explanation = {
            "query": query,
            "search_type": search_type.value,
            "strategy": {
                "description": f"Used {search_type.value} search strategy",
                "components": [],
                "ranking_factors": []
            },
            "results_analysis": {
                "total_results": results.total_results,
                "sources_used": results.sources_used,
                "avg_relevance": sum(r["relevance_score"] for r in results.results) / len(results.results) if results.results else 0
            },
            "performance": {
                "query_time_ms": results.query_time_ms,
                "efficiency_rating": "High" if results.query_time_ms < 500 else "Medium" if results.query_time_ms < 1000 else "Low"
            }
        }
        
        # Add strategy-specific details
        if search_type in [SearchType.VECTOR_ONLY, SearchType.HYBRID]:
            explanation["strategy"]["components"].append("Vector similarity using embeddings")
            explanation["strategy"]["ranking_factors"].append("Cosine similarity score")
        
        if search_type in [SearchType.GRAPH_ONLY, SearchType.HYBRID]:
            explanation["strategy"]["components"].append("Knowledge graph traversal")
            explanation["strategy"]["ranking_factors"].append("Relationship strength")
        
        if search_type == SearchType.TEMPORAL:
            explanation["strategy"]["components"].append("Temporal reasoning")
            explanation["strategy"]["ranking_factors"].append("Temporal relevance")
        
        if search_type == SearchType.PREDICTIVE:
            explanation["strategy"]["components"].append("Pattern analysis")
            explanation["strategy"]["ranking_factors"].append("Prediction confidence")
        
        return explanation
        
    except Exception as error:
        logger.error(f"Search explanation endpoint failed: {error}")
        raise HTTPException(status_code=500, detail=str(error))

# Export the router
__all__ = ["router"]